manager/state/store/extensions.go

```diff
@@ -0,0 +1,272 @@
+package store
+
+import (
+	"errors"
+	"strings"
+
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/manager/state"
+	memdb "github.com/hashicorp/go-memdb"
+)
+
+const tableExtension = "extension"
+
+func init() {
+	register(ObjectStoreConfig{
+		Name: tableExtension,
+		Table: &memdb.TableSchema{
+			Name: tableExtension,
+			Indexes: map[string]*memdb.IndexSchema{
+				indexID: {
+					Name:    indexID,
+					Unique:  true,
+					Indexer: extensionIndexerByID{},
+				},
+				indexName: {
+					Name:    indexName,
+					Unique:  true,
+					Indexer: extensionIndexerByName{},
+				},
+				indexCustom: {
+					Name:         indexCustom,
+					Indexer:      extensionCustomIndexer{},
+					AllowMissing: true,
+				},
+			},
+		},
+		Save: func(tx ReadTx, snapshot *api.StoreSnapshot) error {
+			var err error
+			snapshot.Extensions, err = FindExtensions(tx, All)
+			return err
+		},
+		Restore: func(tx Tx, snapshot *api.StoreSnapshot) error {
+			extensions, err := FindExtensions(tx, All)
+			if err != nil {
+				return err
+			}
+			for _, e := range extensions {
+				if err := DeleteExtension(tx, e.ID); err != nil {
+					return err
+				}
+			}
+			for _, e := range snapshot.Extensions {
+				if err := CreateExtension(tx, e); err != nil {
+					return err
```

Comment:
This is not specific to extensions. Every type of object has similar code. I think changes to how we restore from a snapshot should be part of a separate PR.